### PR TITLE
Bipod now gives you an option to switch to full auto when deploying

### PR DESCRIPTION
**Pull request by VileBeggar**
# About the pull request

Attaching the bipod to weapons with full auto capability or firearms marked with the GUN_SUPPORT_PLATFORM flag will give the user a button toggle to immediately switch to full auto when deploying the bipod.

image

If the full auto switch is turned on, the weapon will simply switch to full auto when the bipod is deployed. If you undeploy it, it will switch back to whatever previous firemode the weapon was in, e.g:

1. You set your pulse rifle to semi auto. 
2. You deploy the bipod. 
3. The pulse rifle is set to full auto. 
4. You undeploy the bipod. 
5. The pulse rifle will switch back to semi auto.

In cases where you switch the weapon's firemode when it's already switched to full auto, it won't revert back to whatever firemode was set before it was deployed, e.g:

1. You set your pulse rifle to semi auto. 
2. You deploy the bipod. 
3. The pulse rifle is set to full auto. 
4. You set your pulse rifle to burst fire.
5. You undeploy the bipod. 
6. The pulse rifle will stay on burst fire.

Also renames a couple of one letter variables.

# Explain why it's good for the game

Quality of life. Less messing around with firemode selectors which is a terrible headache, especially for the HPR, which you usually want to have on full auto 90% of the time. Those who want to stick with the old way don't have to change anything, as the full auto switch is entirely optional.

# Testing Photographs and Procedure
<details>
<summary>Screenshots & Videos</summary>

https://github.com/cmss13-devs/cmss13/assets/17518895/35bcc6e7-947d-4f82-971f-ac22b723537b

</details>

# Changelog
:cl:
qol: Attaching the bipod to weapons that have full auto (HPR included) will give you a button toggle for immediately switching to full auto when deploying the bipod.
/:cl:

### CODE REVIEW

**Comment by cuberound** [comment]
mby you could rename G to gun and A to action to get rid of single letter vars when you are working on that file, would be nice

File: code/modules/projectiles/gun_attachables.dm [hunk 1]
```diff
@@ -3274,14 +3278,31 @@ Defined in conflicts.dm of the #defines folder.
...
 	recoil_mod = RECOIL_AMOUNT_TIER_5
 
-/obj/item/attachable/bipod/Attach(obj/item/weapon/gun/G)
+/obj/item/attachable/bipod/Attach(obj/item/weapon/gun/gun)
 	..()
 
-	RegisterSignal(G, COMSIG_ITEM_DROPPED, PROC_REF(handle_drop))
+	if((GUN_FIREMODE_AUTOMATIC in gun.gun_firemode_list) || (gun.flags_gun_features & GUN_SUPPORT_PLATFORM))
+		var/given_action = FALSE
+		if(usr && (gun == usr.l_hand || gun == usr.r_hand))
```
Comment (harryob) [inline code/modules/projectiles/gun_attachables.dm:3286]: i know we're nullchecking this, but usr is always scary and bad in non-verbs. ideally, we pass an (optional) user var to `/attachable/proc/Attach` from `gun/proc/attach_to_gun` which we can drop in as a replacement for usr here

File: code/modules/projectiles/gun_attachables.dm [hunk 2]
```diff
@@ -3408,6 +3439,31 @@ Defined in conflicts.dm of the #defines folder.
...
+	button.overlays.Cut()
+	button.overlays += image('icons/mob/hud/actions.dmi', button, action_icon_state)
+
+/datum/action/item_action/bipod/toggle_full_auto_switch/action_activate()
+	var/obj/item/weapon/gun/holder_gun = holder_item
+	var/obj/item/attachable/bipod/attached_bipod = holder_gun.attachments["under"]
+
+	attached_bipod.full_auto_switch = !attached_bipod.full_auto_switch
+	to_chat(usr, SPAN_NOTICE("[icon2html(holder_gun, usr)] You will [attached_bipod.full_auto_switch? "<B>start</b>" : "<B>stop</b>"] switching to full auto when deploying the bipod."))
+	playsound(usr, 'sound/weapons/handling/gun_burst_toggle.ogg', 15, 1)
```
Comment (harryob) [inline code/modules/projectiles/gun_attachables.dm:3457]:
```suggestion
	to_chat(owner, SPAN_NOTICE("[icon2html(holder_gun, usr)] You will [attached_bipod.full_auto_switch? "<B>start</b>" : "<B>stop</b>"] switching to full auto when deploying the bipod."))
	playsound(owner, 'sound/weapons/handling/gun_burst_toggle.ogg', 15, 1)
```

**Review by harryob — CHANGES_REQUESTED** [review]

**Review by nauticall — APPROVED** [review]
sprite good 🍋

**Review by Drulikar — CHANGES_REQUESTED** [review]
Minor nitpick, but it seems to only remember the old state early on or when deployed and then gets stuck remembering that state.

1. Be on automatic fire
2. Enable Full Auto Switch
3. Change to burst fire
4. Deploy (changes to automatic)
5. Retract (returns to burst correctly)
6. Change to automatic fire
7. Deploy
8. Retract (incorrectly returns to burst)
9. Change to semi auto
10. Deploy (changes to automatic)
11. Retract (incorrectly returns to burst)

**Review by Drulikar — APPROVED** [review]

**Pull request merged**